### PR TITLE
Move to PyTorch 0.3

### LINKED ISSUE #483: Some tests fail under pytorch 0.3
Hi, 
The folks over at [pytorch](https://github.com/pytorch/pytorch) are working on cutting a new 0.3 release. We'd like to make the transition as smooth as possible so we're testing various community repos that use pytorch with the release.

I built allennlp from source and ran the tests and it looks like the following three failed:
```
tests/training/metrics/entropy_test.py::EntropyTest::test_entropy_for_uniform_distribution 
tests/training/metrics/entropy_test.py::EntropyTest::test_low_entropy_distribution 
tests/training/metrics/entropy_test.py::EntropyTest::test_masked_case 
```
They all seem to be erroring in the same [way](https://gist.github.com/zou3519/0945d32cc77a5ad6f78c3ba131082101). 

cc @soumith

### LINKED ISSUE #513: Move to PyTorch 0.3
We expect PyTorch 0.3 to release on Monday 11/20.

### PR DESCRIPTION
**Pull request by schmmd**
Fixes #483
Fixes #513

### CODE REVIEW

**Comment by schmmd** [comment]
Tests are passing (thanks Mark N) and evaluations for SNLI and BiDAF work fine!  @DeNeutoy was going to tackle the deprecation warnings tomorrow, which are a bit subtle.

**Comment by DeNeutoy** [comment]
@schmmd this is done, I can't get you to review because it's your PR.

File: allennlp/modules/encoder_base.py [hunk 1]
```diff
@@ -173,6 +173,7 @@ def _get_initial_states(self,
             resized_states = []
             # state has shape (num_layers, batch_size, hidden_size)
             for state in self._states:
+                print(state)
```
Comment (matt-gardner) [inline allennlp/modules/encoder_base.py]: Remove print statement.

**Comment by schmmd** [comment]
Thanks @DeNeutoy !  (It ended up being a lot of your PR).

This is a big change, and will break everyone's environment, so I will merge tomorrow morning rather than end of day.

**Comment by DeNeutoy** [comment]
It shouldn't actually break the environment, because all of the changes are backward compatible with pytorch 0.2 and we only install that in the Dockerfile, so pip users will be fine, but merging tomorrow is fine too.

**Comment by schmmd** [comment]
Good point!

**Pull request merged**